gprofiler/metadata/py_package_version.py

```diff
@@ -0,0 +1,258 @@
+            if match is not None:
+                return match.group()
+    return None
+
+
+@functools.lru_cache(maxsize=128)
+def _get_python_full_version(pid: int, short_version: str) -> Optional[str]:
+    assert re.match(r"[23]\.\d\d?", short_version)
+
+    bin_file = _get_libpython_path(pid) or f"/proc/{pid}/exe"
+    full_version_string_pattern = re.compile(rb"(?<=\D)" + short_version.encode() + rb"\.\d\d?(?=\x00)")
+
+    # Try to extract the version string from the binary
+    with open(convert_to_proc_root_path(bin_file, pid), "rb") as f:
+        for line in f.readlines():
+            match = full_version_string_pattern.search(line)
+            if match is not None:
+                return match.group().decode()
+    return None
+
+
+def _get_standard_libs_version(result: Dict[str, Optional[Tuple[str, str]]], pid: int):
+    # Standard library modules are identified by being under a pythonx.y dir and *not* under site/dist-packages
+    standard_lib_pattern = re.compile(r"/python(?P<version>\d\.\d\d?)/(?!.*(site|dist)-packages)")
+    py_version = None
+
+    for path in result:
+        match = standard_lib_pattern.search(path)
+        if match is not None:
+            if py_version is None:
+                try:
+                    py_version = _get_python_full_version(pid, match.group("version"))
+                except OSError:
+                    pass
+                if py_version is None:
+                    # No need to continue trying if we failed
+                    return None
+            result[path] = ("standard-library", py_version)  # type: ignore
+
+
+@functools.lru_cache(maxsize=128)
+def _get_dists_files(packages_path: str) -> Dict[str, pkg_resources.Distribution]:
+    """Return a dict of filename: dist for the distributions in packages_path"""
+    path_to_dist = {}
+    for dist in pkg_resources.find_distributions(packages_path):
+        files_iter = _files_from_record(dist) or _files_from_legacy(dist)
+        if files_iter is not None:
+            path_to_dist.update(dict.fromkeys((os.path.join(packages_path, file) for file in files_iter), dist))
+    return path_to_dist
+
+
+_warned_no__normalized_cached = False
+
+
+def _get_packages_versions(result: Dict[str, Optional[Tuple[str, str]]], pid: int):
+    # A little monkey patch to prevent pkg_resources from converting "/proc/{pid}/root/" to "/".
+    # This function resolves symlinks and makes paths absolute for comparison purposes which isn't required
+    # for our usage.
+    if hasattr(pkg_resources, "_normalize_cached"):
+        original__normalize_cache = pkg_resources._normalize_cached
+        pkg_resources._normalize_cached = lambda path: path
+    else:
+        global _warned_no__normalized_cached
+        if not _warned_no__normalized_cached:
+            # Log only once so we don't spam the log
+            logger.warning("Cannot get modules version, pkg_resources has no '_normalize_cached' attribute")
```

Comment:
Why ever should this scenario happen? We're installing a specific version of setuptools which includes the **exact** `pkg_resources` module we're looking for.

Reply:
Just a safety measure. See [this comment](https://github.com/Granulate/gprofiler/pull/240#discussion_r761832382).